src/ray/object_manager/object_directory.cc

```diff
@@ -156,7 +150,9 @@ ray::Status ObjectDirectory::SubscribeObjectLocations(const UniqueID &callback_i
   // have been evicted from all nodes.
   std::vector<ClientID> client_id_vec(listener_state.current_object_locations.begin(),
                                       listener_state.current_object_locations.end());
-  callback(client_id_vec, object_id);
+  io_service_.post([this, callback, client_id_vec, object_id]() {
+    callback(client_id_vec, object_id);
+  });
```

Comment:
This is the key fix, right?

Reply:
Yes, at least for #3201.

Reply:
The problem is basically the same one that was discussed when `ray.wait` was first implemented [here](https://github.com/ray-project/ray/pull/2162#pullrequestreview-126185059). Basically the callback deletes from a data structure shared with the caller, so an iterator held by the caller gets invalidated when the callback returns.

Reply:
Sorry I'm still confused because I thought we addressed that issue before merging. The iterator in `SubscribeRemainingWaitObjects` makes a copy of the object ids it iterates over ([see here](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/object_manager.cc#L549-L551)), which is not shared with the callback. The callback does not modify the vector of object ids ([see here](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/object_manager.cc#L570-L584)). Could you explain precisely the failure scenario?

Reply:
Yes, but the memory referenced by `wait_state` in that function gets invalidated because the callback deletes the `wait_id` from `active_wait_requests`.

Reply:
If I understand correctly, one failure scenario is as follows:
1. The final object id's callback is invoked immediately.
2. The `wait_id` is removed within the callback (i.e. `WaitComplete` is invoked).
3. At this point, the loop exits, so [this](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/object_manager.cc#L559-L564) is not invoked.
4. [This](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/object_manager.cc#L587) is invoked, which references memory that doesn't exist.

Reply:
Yes.